### PR TITLE
refactor: exactOptionalPropertyTypes

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Context
- Ref: https://github.com/renovatebot/renovate/issues/20187

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by JamieMagee — DISMISSED** [review]
Are there more changes required, or can you enable `exactOptionalPropertyTypes` now?

**Comment by RahulGautamSingh** [comment]
> Are there more changes required, or can you enable `exactOptionalPropertyTypes` now?

Many more :) There are too many changes so I am dividing them across PRs.

File: lib/modules/datasource/galaxy-collection/index.ts [hunk 1]
```diff
@@ -93,7 +93,7 @@ export class GalaxyCollectionDatasource extends Datasource {
             try {
               const release: Release = {
                 version: basicRelease.version,
-                isDeprecated: basicRelease.isDeprecated,
+                isDeprecated: Boolean(basicRelease.isDeprecated),
```
Comment (viceice) [inline lib/modules/datasource/galaxy-collection/index.ts:96]:
```suggestion
                isDeprecated: !!basicRelease.isDeprecated,
```

File: lib/modules/datasource/docker/index.ts [hunk 2]
```diff
@@ -234,7 +234,7 @@ export async function getAuthHeaders(
 }
 
 async function getECRAuthToken(
-  region: string | undefined,
+  region: string,
```
Comment (viceice) [inline lib/modules/datasource/docker/index.ts:237]: are we sure region is never undefined here?

File: lib/workers/repository/update/pr/index.ts [hunk 3]
```diff
@@ -47,11 +47,13 @@ export function getPlatformPrOptions(
...
-    bbUseDefaultReviewers: config.bbUseDefaultReviewers,
-    gitLabIgnoreApprovals: config.gitLabIgnoreApprovals,
-    forkModeDisallowMaintainerEdits: config.forkModeDisallowMaintainerEdits,
+    azureAutoApprove: Boolean(config.azureAutoApprove),
+    azureWorkItemId: Number(config.azureWorkItemId),
+    bbUseDefaultReviewers: Boolean(config.bbUseDefaultReviewers),
+    gitLabIgnoreApprovals: Boolean(config.gitLabIgnoreApprovals),
+    forkModeDisallowMaintainerEdits: Boolean(
+      config.forkModeDisallowMaintainerEdits
+    ),
```
Comment (viceice) [inline lib/workers/repository/update/pr/index.ts:56]: you should fix the types instead

File: lib/workers/repository/update/pr/index.ts [hunk 4]
```diff
@@ -385,7 +387,7 @@ export async function ensurePr(
           prBody,
           labels: prepareLabels(config),
           platformOptions: getPlatformPrOptions(config),
-          draftPR: config.draftPR,
+          draftPR: Boolean(config.draftPR),
```
Comment (viceice) [inline lib/workers/repository/update/pr/index.ts:390]:
```suggestion
          draftPR: !!config.draftPR,
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/docker/index.ts (hunk 2 above)
Comment (RahulGautamSingh) [reply lib/modules/datasource/docker/index.ts:237]: https://github.com/renovatebot/renovate/blob/1f22ed2a0e4272e285358632e43b85d3799efa98/lib/modules/datasource/docker/index.ts#L105-L115

https://github.com/renovatebot/renovate/blob/1f22ed2a0e4272e285358632e43b85d3799efa98/lib/modules/datasource/docker/index.ts#L52

I think the regex test is sufficient for that.

File: lib/modules/platform/util.ts [hunk 5]
```diff
@@ -15,3 +16,12 @@ export function getNewBranchName(branchName?: string): string | undefined {
...
 }
+
+export function getDefaultPlatformPrOptions(): PlatformPrOptions {
+  return {
+    azureAutoApprove: false,
+    azureWorkItemId: 0,
+    bbUseDefaultReviewers: false,
+    gitLabIgnoreApprovals: 0,
+  };
+}
```
Comment (RahulGautamSingh) [inline lib/modules/platform/util.ts:27]: @viceice Do you think this is a good idea?

File: lib/modules/platform/types.ts [hunk 6]
```diff
@@ -87,13 +87,14 @@ export interface Issue {
...
 }
 export type PlatformPrOptions = {
-  azureAutoApprove?: boolean;
-  azureWorkItemId?: number;
-  bbUseDefaultReviewers?: boolean;
-  gitLabIgnoreApprovals?: boolean;
+  azureAutoApprove: boolean;
+  azureWorkItemId: number;
+  bbUseDefaultReviewers: boolean;
+  gitLabIgnoreApprovals: number;
```
Comment (viceice) [inline lib/modules/platform/types.ts:93]: ? why number? shouldn't we convert the bool option to corresponding number value at platform code?

File: lib/modules/platform/util.ts (hunk 5 above)
Comment (viceice) [reply lib/modules/platform/util.ts:27]: no, should be handled inside getplatformoptions

File: lib/modules/platform/types.ts (hunk 6 above)
Comment (viceice) [reply lib/modules/platform/types.ts:93]: i think they should stay optional and the platform code should handle it

File: lib/workers/repository/update/pr/index.ts [hunk 7]
```diff
@@ -47,11 +47,11 @@ export function getPlatformPrOptions(
   );
 
   return {
-    azureAutoApprove: config.azureAutoApprove,
-    azureWorkItemId: config.azureWorkItemId,
-    bbUseDefaultReviewers: config.bbUseDefaultReviewers,
-    gitLabIgnoreApprovals: config.gitLabIgnoreApprovals,
-    forkModeDisallowMaintainerEdits: config.forkModeDisallowMaintainerEdits,
+    azureAutoApprove: !!config.azureAutoApprove,
```
Comment (viceice) [inline lib/workers/repository/update/pr/index.ts:50]: we should make the target type the same in a follow up PR and platform code should handle undefined and null values.

**Review by viceice — APPROVED** [review]

**Pull request merged**